gptel-gh: Change how tokens are managed

marcus-lundgren opened the pull request:
Introduce functions that can be overloaded that controls how a token is restored/stored. This enables the user to make gptel-gh store the token in a secure manner, similar to what is recommended for the other backend variants.

In conjunction with this, proper support for multiple accounts has also been introduced by having each gptel-gh backend be marked with which account they should be considered to belong to.


**Old description:**
Removed the file based persistent storage of the tokens, as they were stored in clear text. They will only be stored in variables.

The initial value of the github-token can now be set by providing a function that provides it. This enables this token to be securely stored outside of Emacs.

marcus-lundgren committed:
gptel-gh: Change how tokens are managed

Introduced overloadable function variables that determines how a token
is loaded and saved.

gptel-gh: Introduced logic for backends to share the github-token

This commit introduces proper support for more than one GitHub
account. By stating which account a backend should be considered to be
connected to, it is possible to ensure that we don't e.g. login using
a personal account to a backend considered for the work account.

The file based persistence will make use of the stated GitHub username
and make that a suffix to the configured file path. As the username
for GitHub accounts may only contain [0-9a-zA-Z], they can safely be
used for filenames.

The custom save/load function now takes the username as a parameter
so that it can decide on how to proceed.

marcus-lundgren committed:
gptel-oauth: Change how tokens are managed

Introduce account-hint concept in gptel-openai-oauth and gptel-gh
and custom load/save function for oauth.

marcus-lundgren committed:
gptel-gh: Removed unnecessary call to (current-time)

(float-time) will implicitly use (current-time) if the argument is not specified.

marcus-lundgren committed:
gptel-gh: Removed unnecessary call to (current-time)

(float-time) will implicitly use (current-time) if the argument is not specified.

marcus-lundgren commented:
@karthink - Do you have any input in how you imagine the workflow to be for this API regarding the GitHub token?

With the changes made just recently to present the login function, is it reasonable to have a workflow to perform that in order to get the token that can then be stored elsewhere?

karthink commented:
> @karthink - Do you have any input in how you imagine the workflow to be for this API regarding the GitHub token?
>
> With the changes made just recently to present the login function, is it reasonable to have a workflow to perform that in order to get the token that can then be stored elsewhere?

Sorry, I haven't looked at this PR yet.  I also don't use Copilot so let's hear from @kiennq (the author of gptel-gh) on the proposed changes.

marcus-lundgren commented:
@kiennq  - A friendly reminder about this PR :)

marcus-lundgren commented:
@kiennq, @karthink - another reminder.

karthink commented:
@marcus-lundgren just dropping a note to say this PR remains on my radar -- I just need a longer chunk of time to review it than other PRs since I'm not the author of the token management code.  I'll do a general code review but I trust @kiennq's judgment on the design.

kiennq commented on gptel-gh.el:
```diff
@@ -317,38 +302,25 @@ If your browser does not open automatically, browse to %s."
-    ;; try to load token from `gptel-gh-token-file'
-    (setf (gptel--gh-token gptel-backend)
-          (gptel--gh-restore gptel-gh-token-file)))
+  (unless (funcall gptel-gh-github-token-load)
```
This function can be called quite frequently. Can you change the default `token-load` function to cache github-token to a variable instead of reading from file everytime?

marcus-lundgren replied:
Good catch! I introduced a general cache for the loading function. This will also mimic the behavior from when the token was stored in the backend.

marcus-lundgren commented on gptel-gh.el:
```diff
@@ -180,26 +180,28 @@
-  "File where the chat token is cached."
-  :type 'string
-  :group 'gptel)
+(defvar gptel-gh-github-token-load 'gptel--gh-restore-from-file
```
Fixed! When I investigated defcustom previously, I for some reason was unable to find the function type.

marcus-lundgren commented:
Updated config example with the builtin cache functionality:

```emacs-lisp
(use-package gptel
  :defer t
  :config
  (setq
   gptel-gh-github-token-load-function (lambda () (read-passwd "Copilot token: "))
   gptel-gh-github-token-save-function (lambda (token) (message "Token saved: %s" token))
   gptel-backend (gptel-make-gh-copilot "Copilot"))
  )

```

kiennq commented on gptel-gh.el:
```diff
@@ -287,28 +291,24 @@ in your local browser, enter the code, and authorize.  Press ENTER after authori
-        (message "Successfully logged in to GitHub Copilot.")
-      (user-error "Error: You might not have access to GitHub Copilot Chat!"))))
+             user_code verification_uri))
+    (browse-url verification_uri)
```
The indentation looks off, can you double check?

marcus-lundgren commented:
@kiennq, @karthink - I resolved the review comments you had after fixing them. If the workflow is that the reviewer should do that, then please let me know.

I've now also updated the README file with information about the customizable variables.

Some suitable information for the NEWS file can be found below. Do you want me to add it?

- Breaking change: Removed the customizable variable `gptel-gh-token-file`, as the chat token is no longer persistently stored.
- Added the customizable variables `gptel-gh-github-token-load-function` and `gptel-gh-github-token-save-function`. This enables the GitHub OAuth token to be loaded from and saved to arbitrary places.

karthink commented:
If I understand the changes correctly, you can no longer have more than one Github copilot backend defined, since the authentication is now shared across all gh backends?  Is this correct?

marcus-lundgren commented:
@karthink - Yes, you are correct, but I would argue that while it was technically true before, it wasn't properly implemented as such. Simply due to the fact that one could only persistently store one token and there was no official way to set the GitHub token manually.

Calling the login function will overwrite the token stored in the file:

https://github.com/karthink/gptel/blob/master/gptel-gh.el#L267

It would also fall back to the persistently stored token on failure to renew the chat token:

https://github.com/karthink/gptel/blob/master/gptel-gh.el#L287 followed by
https://github.com/karthink/gptel/blob/master/gptel-gh.el#L298

With the changes I've made as a basis, one could:

- Reintroduce the tokens to the backend to be used as a cache variable
- Add the load and save functions as parameters to the gptel-make-gh-copilot function and add them as references to the backend. This would enable several backends to use the same overloaded functions.

But I think that might be suitable for a follow up PR to make it proper.

Any thoughts, @kiennq? Do you agree with my understanding of the logic and that it is suitable for another PR?

kiennq commented on gptel-gh.el:
```diff
@@ -182,25 +199,25 @@
-(defun gptel--gh-restore (file)
-  "Restore saved object from FILE."
-  (when (file-exists-p file)
+(defun gptel--gh-restore-from-file ()
```
Can you keep this function the same signature as before, i.e., taking a file param.
I'm working on a change to automatically fetch the model list and will use this to store the cached model list as well

marcus-lundgren commented:
I am still determined to follow through with this PR, but family and other things has to be prioritized at the moment. I am still using this branch every day at work, so it will not become stale.

kiennq commented on gptel-gh.el:
```diff
@@ -411,6 +464,7 @@ for."
                   :stream stream
                   :request-params request-params
                   :curl-args curl-args
+                  :github-username github-username
```
Can we use the `:name` parameter instead? I'm not sure if adding new parameter is better since we need unique `name` for each backend anyway.

marcus-lundgren replied:
We could, but then:

- We would not be able to have the current implementation's benefit of updated all backends that are connected to the same account. But if the officially sanctioned logic is that one GitHub account should only be connected to one backend, then I can remove that logic no matter which solution we choose.
- How should illegal characters for filenames be handled? The `:github-username` parameter enforces the allowed characters for GitHub usernames, which fits within the valid character range for filenames. The `:name` parameter may currently take any character, such as `/`, which is problematic. A solution would be to add the enforcement to that parameter as well, but would confusingly only be a requirement for the copilot backend.

Even though it is unfortunate, I feel that having this separate from the `:name` parameter is the least bad choice.

kiennq commented:
Approve with a suggestion of not adding `:github-username` parameter for the class. We have `:name` for each backend already

kiennq commented on gptel-gh.el:
```diff
@@ -219,32 +284,26 @@
     obj))
 
-(defun gptel-gh-login ()
+(defun gptel-gh-login (github-username)
```
This function is interactive, the `github-username` should be able to be inputted from mini-buffer when calling this interactively

marcus-lundgren replied:
Good catch! I'll fix it.

kiennq replied:
Btw, here is how I fixed it in my branch. Your changes are working fine for me though. https://github.com/kiennq/gptel/commit/8e0bb5f1523802959c6c3860d3665717551b68b9

kiennq commented on gptel-gh.el:
```diff
@@ -230,32 +310,37 @@
+                       (cond
+                        ((= 0 (length known-usernames))
+                         (user-error "No GitHub copilot backends registered"))
+                        ((= 1 (length known-usernames))
```
In this case if we only have default user, it would return the place holder and then we can't find that placeholder on the backend list

marcus-lundgren commented:
@karthink, @kiennq - A friendly reminder regarding this PR :-)

marcus-lundgren commented:
@karthink - Are you looking at the latest version? I have reverted from the global gptel--gh-token so that it is placed in each backend as before.

karthink commented:
Haven't lost track of this PR!  It's fast approaching the title of longest standing gptel PR that will eventually be merged.

Incidentally, see discussion of at-rest oauth token encryption in #1349.

marcus-lundgren commented:
> Haven't lost track of this PR! It's fast approaching the title of longest standing gptel PR that will eventually be merged.

Thanks for the update, @karthink! I will wear it as a badge of honor if that becomes the case :-)

> Incidentally, see discussion of at-rest oauth token encryption in #1349.

Yeah, I saw it whilst performing my daily browsing of issues and PRs in this repo. It is always nice when someone is interested in a thing that one has created (which I hope is a feeling you get for having created this wonderful package).

marcus-lundgren commented:
Thank you for the update, @karthink.

I saw that it was merged and I have been working for some time on rebasing my changes to latest master this evening. In order to reduce the complexity of it, I've squashed some of the commits together. I don't have much time at the moment so it may take some time, but it will be done in the end.

I will first start with getting the branch in a state where the external functionality is the same. I will then proceed with making the custom functions to be placed in the new file. Will let you know as soon I have a need to discuss some implementation detail.

marcus-lundgren committed:
Fixed indentation

marcus-lundgren committed:
gptel-gh: Fixed typo

marcus-lundgren committed:
Updated README

marcus-lundgren committed:
Introduce backend-type parameter to the custom load/save functions

This allows the custom function to know which backend the account-hint
belongs to. A highly likely case is the "default" account that is the
empty string.

marcus-lundgren committed:
Fix openai-oauth loading of plist

marcus-lundgren committed:
Update custom function examples in the README

marcus-lundgren committed:
Fix missing closing parenthesis

marcus-lundgren committed:
Remove redundant code

marcus-lundgren committed:
Remove redundant code

marcus-lundgren committed:
Remove redundant code

marcus-lundgren committed:
Fix missing closing parenthesis

marcus-lundgren committed:
Validate the token when saving/loading

marcus-lundgren committed:
Removed mention of plist in the documentation of load/save functions

The type of the token is unknown to gptel-oauth.

marcus-lundgren committed:
Fixed indendation

marcus-lundgren committed:
Always propagate the loaded token

marcus-lundgren committed:
Fixed missing funcall

marcus-lundgren committed:
Fixed check if custom functions are defined (not nil)

marcus-lundgren committed:
Fix markup

marcus-lundgren commented:
@karthink - A gentle bump regarding this PR.
